packages/svelte-ux/src/lib/styles/theme.ts

```diff
@@ -65,43 +65,50 @@ export function getThemeNames(themes: Record<string, any>) {
   return { light, dark };
 }
 
+export interface NestedColors {
+  [key: string]: string | NestedColors;
+}
+
 /**
- * Convert names to CSS variables and color values common color space (hsl, oklch, etc) and space separated
+ * Flatten nested color objects into a single-level color object with concatenated keys
  */
-export function processThemeColors(
-  themeColors: Record<string, string>,
-  colorSpace: SupportedColorSpace
+export function flattenColors(
+  themeColors: NestedColors,
+  keys: (string | number)[] = [],
+  memo?: Record<string, string>
 ) {
-  const colors = { ...themeColors };
+  return entries(themeColors).reduce<Record<string, string>>((memo, [key, value]) => {
+    const newKeys = [...keys, key];
+    if (typeof value === 'string') {
+      memo[newKeys.join('-')] = value;
+    } else {
+      flattenColors(value, newKeys, memo);
```

Comment:
Is there a need for this to be recursive (can you provide an example)?

If not, I would like to simplify this to something like...
```js
return fromEntries(entries(colors).flatMap(([color, value]) => {
	if (typeof value === 'string') {
		return [color, value];
	} else {
		return entries(value).map(([shade, value]) => {
			return [`${color}-${shade}`, value]
		})
	}
}))
```

Reply:
I use a similar optimization in my active PR:
![](https://github.com/techniq/svelte-ux/assets/5913254/b6018e6d-fc87-49cb-b3af-304e430ffb96)

I added support for recursion so if anyone wants to pass in an object that contains multiple levels, they don't have to do the flattening themselves, e.g.:
```ts
const colors = {
  gray: {
    cool: {
      50: '#f8fafc',
      100: '#f1f5f9',
      200: '#e2e8f0',
      300: '#cbd5e1',
      400: '#94a3b8',
      500: '#64748b',
      600: '#475569',
      700: '#334155',
      800: '#1e293b',
      900: '#0f172a',
      950: '#020617',
    },
    warm: {
      50: '#fafaf9',
      100: '#f5f5f4',
      200: '#e7e5e4',
      300: '#d6d3d1',
      400: '#a8a29e',
      500: '#78716c',
      600: '#57534e',
      700: '#44403c',
      800: '#292524',
      900: '#1c1917',
      950: '#0c0a09',
    },
  }
}
```

This same idea is also actively being discussed for inclusion in CSS (primarily for use in CSS libs like Tailwind CSS) here:
https://github.com/w3c/csswg-drafts/issues/9992

Reply:
@techniq If you would still rather this only support one level, I can refactor this. Just lmk

Reply:
I'm fine with the recursion.  A lot of the times I think of `-content` as it's own color and reference it explicitly (`primary-content`, etc) but supporting it in the same object is nice.

Care to add a test for `flattenColors()` (create `theme.test.ts` ) to capture these use cases?  I need to add tests for more of the theme functions (`processThemeColors()`, etc), but this would be a good and simple start.  Thanks!

Reply:
Tests added for both `flattenColors` and `processThemeColors`, and support added for `'DEFAULT'` shade, matching Tailwind's nomenclature, though this will be usable by any CSS library.